aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java

```diff
@@ -0,0 +1,173 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.aws.s3;
+
+import com.amazonaws.services.s3.AmazonS3URI;
+import java.io.IOException;
+import java.io.Serializable;
+import org.apache.http.HttpStatus;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.io.OutputFile;
+import org.apache.iceberg.io.PositionOutputStream;
+import org.apache.iceberg.io.SeekableInputStream;
+import org.apache.iceberg.util.SerializableSupplier;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.Delete;
+import software.amazon.awssdk.services.s3.model.DeleteObjectsRequest;
+import software.amazon.awssdk.services.s3.model.HeadObjectRequest;
+import software.amazon.awssdk.services.s3.model.HeadObjectResponse;
+import software.amazon.awssdk.services.s3.model.ObjectIdentifier;
+import software.amazon.awssdk.services.s3.model.S3Exception;
+
+/**
+ * FileIO implementation backed by S3.  Locations used must follow the conventions for URIs (e.g. s3://bucket/path/..).
+ */
+public class S3FileIO implements FileIO, Serializable {
+  private final SerializableSupplier<S3Client> s3;
+  private transient S3Client client;
+
+  public S3FileIO(SerializableSupplier<S3Client> s3) {
```

Comment:
What is the reasoning behind having a serializable S3Client supplier? Since you are already marking the s3 `client` as transient, why not just have:

```java
public S3FileIO() {
  client = S3Client.create();
}
```

Also the AWS client is hard to serialize. You need to serialize things like credential information that might expire after encode and decode. It seems hard to use this constructor unless we also provide an actual implementation of `SerializableSupplier<S3Client`.

Reply:
The `SerializableSupplier` is there because the client is hard to serialize, as you mentioned. The supplier can carry information needed to create a client and is itself serializable instead.

We need to serialize the `FileIO` because it is sent to tasks in Spark and Flink. If the `FileIO` can be sent but then can't instantiate a configured client, then it won't work as intended.

Reply:
Sorry I wasn't clear in the comment, I understand the serialization aspect of this, and what I am proposing is to have a default behavior that does not serialize the client, and simply recreate it if it is null:

```java
    private final SerializableSupplier<S3Client> s3;
    private transient S3Client client;

    public S3FileIO() {
        this.s3 = null;
    }

    public S3FileIO(SerializableSupplier<S3Client> s3) {
        this.s3 = s3;
    }

    private S3Client client() {
        if (client == null) {
            if (s3 != null) {
                client = s3.get();
            } else {
                client = S3Client.create();
            }
        }
        return client;
    }
```
The reason is because the environment that deserializes the object is very likely to have the AWS configurations for the client to pick up and re-initialize itself. This is the situation for all AWS compute including EC2, EMR, ECS, Lambda. By having this default, most users can just use it as `new S3FileIO()` without the need to implement a serialization supplier.

The only situation left is the users with customized components such as a customized credential provider or HTTP client. In those cases, the current constructor can be used with a customized supplier.

Reply:
It makes sense to me to have a constructor that doesn't require the client supplier.

Reply:
I'll add the default constructor.